aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/RequestRetry.kt

```diff
@@ -0,0 +1,45 @@
+package com.amplifyframework.datastore.syncengine
+
+import com.amplifyframework.AmplifyException
+
+import io.reactivex.rxjava3.core.Single
+import io.reactivex.rxjava3.core.SingleEmitter
+import java.util.concurrent.TimeUnit
+import kotlin.math.pow
+
+
+class RequestRetry(private val maxExponent: Int = 8,
+                    private val jitterFactor: Int = 100,
+                    private val maxAttempts: Int = 3) {
+
+    private var numberOfAttempts = 0
+
+
+        fun <T> retry(single: Single<T>, skipExceptions: List<Class<out Throwable>?>): Single<T> {
+            return Single.create { emitter -> call(single, emitter, 0, skipExceptions) }
+        }
+
+        private fun <T> call(single: Single<T>, emitter: SingleEmitter<T>, delayInSeconds: Long, skipExceptions: List<Class<out Throwable>?> ) {
+           single.delaySubscription(delayInSeconds, TimeUnit.SECONDS)
+                .subscribe({
+                            t: T -> emitter.onSuccess(t)
+                }) { error ->
+                    numberOfAttempts++
+                    if (numberOfAttempts > maxAttempts || (error.javaClass in skipExceptions)) {
```

Comment:
FYI, we have a `ErrorInspector` utility class that checks to see if a `Throwable` is a certain type, or if any nested `cause` is of that type.  I'm not entirely sure if you need to check causes, but I'm guessing you might, since you are probably going to have a `DataStoreException` here, which has a cause that is an `NonRetryableException`.

Reply:
Done